recipes/openai_ab.py

```diff
@@ -0,0 +1,365 @@
+    prompts: Dict[str, jinja2.Template]
+    inputs: Iterable[PromptInput]
+    batch_size: int
+    verbose: bool
+    randomize: bool
+    openai_api_org: str
+    openai_api_key: str
+    openai_temperature: float
+    openai_max_tokens: int
+    openai_timeout_s: int
+    openai_n: int
+
+    def __init__(
+        self,
+        display: jinja2.Template,
+        prompts: Dict[str, jinja2.Template],
+        inputs: Iterable[PromptInput],
+        *,
+        openai_api_org: str,
+        openai_api_key: str,
+        openai_model: str,
+        batch_size: int = 10,
+        verbose: bool = False,
+        randomize: bool = True,
+        openai_temperature: float = 0,
+        openai_max_tokens: int = 500,
+        openai_timeout_s: int = 50,
+        openai_n: int = 1,
+    ):
+        self.display = display
+        self.inputs = inputs
+        self.prompts = prompts
+        self.model = openai_model
+        self.batch_size = batch_size
+        self.verbose = verbose
+        self.openai_api_org = openai_api_org
+        self.openai_api_key = openai_api_key
+        self.openai_temperature = openai_temperature
+        self.openai_max_tokens = openai_max_tokens
+        self.openai_timeout_s = openai_timeout_s
+        self.openai_n = openai_n
+        self.randomize = randomize
+
+    def __iter__(self) -> Iterable[Dict]:
+        for input_batch in _batch_sequence(self.inputs, self.batch_size):
+            response_batch = self._get_response_batch(input_batch)
+            for input_, responses in zip(input_batch, response_batch):
+                yield self._make_example(
+                    input_.id,
+                    self.display.render(**input_.prompt_args),
+                    responses,
+                    randomize=self.randomize,
+                )
+
+    def progress(self) -> float:
+        return 0.0
+
+    def on_exit(self, ctrl):
+        examples = ctrl.db.get_dataset_examples(ctrl.dataset)
+        counts = Counter()
+        # Get last example per ID
+        for eg in examples:
+            selected = eg.get("accept", [])
+            if not selected or len(selected) != 1 or eg["answer"] != "accept":
+                continue
+            counts[selected[0]] += 1
+        print("")
+        if not counts:
+            msg.warn("No answers found", exits=0)
+        msg.divider("Evaluation results", icon="emoji")
+        pref, _ = counts.most_common(1)[0]
+        msg.good(f"You preferred {pref}")
+        rows = [(name, count) for name, count in counts.most_common()]
+        msg.table(rows, aligns=("l", "r"))
+
+    def _get_response_batch(self, inputs: List[PromptInput]) -> List[Dict[str, str]]:
+        name1, name2 = self._choose_rivals()
+        prompts = []
+        for input_ in inputs:
+            prompts.append(self._get_prompt(name1, input_.prompt_args))
+            prompts.append(self._get_prompt(name2, input_.prompt_args))
+        responses = self._get_responses(prompts)
```

Comment:
It might be nice to show the actual queries on the CLI when the `verbose` flag is used:
```suggestion
        if self.verbose:
            for prompt in prompts:
                rich.print(Panel(prompt, title="Prompt to OpenAI"))
        responses = self._get_responses(prompts)
```
(requires adding the correct imports, too)